src/server/pps/server/master.go

```diff
@@ -151,129 +174,47 @@ eventLoop:
 	for {
 		select {
 		case e := <-m.eventCh:
-			switch e.eventType {
-			case writeEv:
-				if err := m.attemptStep(m.masterCtx, e); err != nil {
-					log.Errorf("PPS master: %v", err)
+			func(e *pipelineEvent) {
+				m.opsInProcessMu.Lock()
+				defer m.opsInProcessMu.Unlock()
+
+				var ot opType
+				switch e.eventType {
+				case writeEv:
+					ot = writeOp // raises flag in pipelineOp to run again whenever it finishes (doesn't block)
+				case deleteEv:
+					ot = deleteOp
 				}
-			case deleteEv:
-				// TODO(msteffen) trace this call
-				if err := m.deletePipelineResources(e.pipeline); err != nil {
-					log.Errorf("PPS master: could not delete resources for pipeline %q: %v",
-						e.pipeline, err)
+
+				if pipelineOp, ok := m.opsInProcess[e.pipeline]; ok {
+					pipelineOp.Bump(ot) // acquires lock: pipelineOp.bumpLock
```

Comment:
Bump internally acquires a lock making it blocking. Bump is slightly more complicated than I originally suspected because it also has to account for a delete operation, as those should also be serialized within a Pipeline Operation goroutine

@msteffen in our discussion, you had mentioned that Ideally this could be done as a non-blocking call. Do you think that's still possible here?